### PR TITLE
feat(es/lexer): Add TS1185 error

### PR DESCRIPTION
**Pull request by magic-akari**
**Description:**

**BREAKING CHANGE:**

**Related issue (if exists):**

### CODE REVIEW

**Review by kdy1 — COMMENTED** [review]
Do we need this?
I expect this to have negative impact on performance even for valid inputs

**Comment by magic-akari** [comment]
The detection logic is in a deeper branch, so I expect it to have limited negative impact.

It will only call `eat_str` if all of the following conditions are met.
- after line break
- `<<` | `>>>` | `===` | `||`

I would measure the performance impact.

**Comment by magic-akari** [comment]
### main(bad9811)

```
test es/lexer/colors        ... bench:        9587 ns/iter (+/- 49)
test es/lexer/angular       ... bench:     4762035 ns/iter (+/- 10865)
test es/lexer/backbone      ... bench:      660254 ns/iter (+/- 1834)
test es/lexer/jquery        ... bench:     3550302 ns/iter (+/- 8351)
test es/lexer/jquery mobile ... bench:     5708467 ns/iter (+/- 13252)
test es/lexer/mootools      ... bench:     2755662 ns/iter (+/- 5354)
test es/lexer/underscore    ... bench:      546400 ns/iter (+/- 901)
test es/lexer/three         ... bench:    16100536 ns/iter (+/- 31573)
test es/lexer/yui           ... bench:     2934677 ns/iter (+/- 5869)
```

### After 57a4992

```
test es/lexer/colors        ... bench:        9514 ns/iter (+/- 49)
test es/lexer/angular       ... bench:     4748553 ns/iter (+/- 11199)
test es/lexer/backbone      ... bench:      654384 ns/iter (+/- 1700)
test es/lexer/jquery        ... bench:     3536403 ns/iter (+/- 9467)
test es/lexer/jquery mobile ... bench:     5681606 ns/iter (+/- 12360)
test es/lexer/mootools      ... bench:     2746125 ns/iter (+/- 6172)
test es/lexer/underscore    ... bench:      544793 ns/iter (+/- 1487)
test es/lexer/three         ... bench:    16019182 ns/iter (+/- 26406)
test es/lexer/yui           ... bench:     2924109 ns/iter (+/- 5955)
```

**Comment by kdy1** [comment]
You should close all github PR pages or other background programs while running benchmark. I'll try it

**Comment by kdy1** [comment]
I'll update this soon

## main

```
Gnuplot not found, using plotters backend
Benchmarking es/parser/colors
Benchmarking es/parser/colors: Warming up for 3.0000 s
Benchmarking es/parser/colors: Collecting 100 samples in estimated 5.0726 s (318k iterations)
Benchmarking es/parser/colors: Analyzing
es/parser/colors        time:   [15.766 us 15.788 us 15.815 us]
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe

Benchmarking es/parser/angular
Benchmarking es/parser/angular: Warming up for 3.0000 s
Benchmarking es/parser/angular: Collecting 100 samples in estimated 5.0474 s (600 iterations)
Benchmarking es/parser/angular: Analyzing
es/parser/angular       time:   [8.1861 ms 8.2031 ms 8.2220 ms]
Found 12 outliers among 100 measurements (12.00%)
  5 (5.00%) high mild
  7 (7.00%) high severe

Benchmarking es/parser/backbone
Benchmarking es/parser/backbone: Warming up for 3.0000 s

Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.8s, enable flat sampling, or reduce sample count to 60.
Benchmarking es/parser/backbone: Collecting 100 samples in estimated 5.7935 s (5050 iterations)
Benchmarking es/parser/backbone: Analyzing
es/parser/backbone      time:   [1.1399 ms 1.1416 ms 1.1433 ms]
Found 12 outliers among 100 measurements (12.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  8 (8.00%) high severe

Benchmarking es/parser/jquery
Benchmarking es/parser/jquery: Warming up for 3.0000 s
Benchmarking es/parser/jquery: Collecting 100 samples in estimated 5.1112 s (800 iterations)
Benchmarking es/parser/jquery: Analyzing
es/parser/jquery        time:   [6.3346 ms 6.3475 ms 6.3620 ms]
Found 17 outliers among 100 measurements (17.00%)
  4 (4.00%) high mild
  13 (13.00%) high severe

Benchmarking es/parser/jquery mobile
Benchmarking es/parser/jquery mobile: Warming up for 3.0000 s
Benchmarking es/parser/jquery mobile: Collecting 100 samples in estimated 5.0423 s (500 iterations)
Benchmarking es/parser/jquery mobile: Analyzing
es/parser/jquery mobile time:   [9.9538 ms 9.9748 ms 9.9982 ms]
Found 15 outliers among 100 measurements (15.00%)
  4 (4.00%) high mild
  11 (11.00%) high severe

Benchmarking es/parser/mootools
Benchmarking es/parser/mootools: Warming up for 3.0000 s
Benchmarking es/parser/mootools: Collecting 100 samples in estimated 5.3064 s (1100 iterations)
Benchmarking es/parser/mootools: Analyzing
es/parser/mootools      time:   [4.8053 ms 4.8142 ms 4.8241 ms]
Found 13 outliers among 100 measurements (13.00%)
  4 (4.00%) high mild
  9 (9.00%) high severe

Benchmarking es/parser/underscore
Benchmarking es/parser/underscore: Warming up for 3.0000 s
Benchmarking es/parser/underscore: Collecting 100 samples in estimated 9.8987 s (10k iterations)
Benchmarking es/parser/underscore: Analyzing
es/parser/underscore    time:   [976.25 us 977.17 us 978.19 us]
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe

Benchmarking es/parser/three
Benchmarking es/parser/three: Warming up for 3.0000 s
Benchmarking es/parser/three: Collecting 100 samples in estimated 5.6801 s (200 iterations)
Benchmarking es/parser/three: Analyzing
es/parser/three         time:   [28.197 ms 28.275 ms 28.360 ms]
Found 9 outliers among 100 measurements (9.00%)
  7 (7.00%) high mild
  2 (2.00%) high severe

Benchmarking es/parser/yui
Benchmarking es/parser/yui: Warming up for 3.0000 s
Benchmarking es/parser/yui: Collecting 100 samples in estimated 5.3964 s (1100 iterations)
Benchmarking es/parser/yui: Analyzing
es/parser/yui           time:   [4.8567 ms 4.8662 ms 4.8769 ms]
Found 14 outliers among 100 measurements (14.00%)
  5 (5.00%) high mild
  9 (9.00%) high severe
```

**Comment by kdy1** [comment]
```
Gnuplot not found, using plotters backend
Benchmarking es/parser/colors
Benchmarking es/parser/colors: Warming up for 3.0000 s
Benchmarking es/parser/colors: Collecting 100 samples in estimated 5.0593 s (318k iterations)
Benchmarking es/parser/colors: Analyzing
es/parser/colors        time:   [15.783 us 15.825 us 15.880 us]
                        change: [-0.4053% +0.0079% +0.4486%] (p = 0.97 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  4 (4.00%) high severe

Benchmarking es/parser/angular
Benchmarking es/parser/angular: Warming up for 3.0000 s
Benchmarking es/parser/angular: Collecting 100 samples in estimated 5.0044 s (600 iterations)
Benchmarking es/parser/angular: Analyzing
es/parser/angular       time:   [8.1345 ms 8.1497 ms 8.1671 ms]
                        change: [-0.9462% -0.6515% -0.3579%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 17 outliers among 100 measurements (17.00%)
  3 (3.00%) high mild
  14 (14.00%) high severe

Benchmarking es/parser/backbone
Benchmarking es/parser/backbone: Warming up for 3.0000 s

Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.8s, enable flat sampling, or reduce sample count to 60.
Benchmarking es/parser/backbone: Collecting 100 samples in estimated 5.7823 s (5050 iterations)
Benchmarking es/parser/backbone: Analyzing
es/parser/backbone      time:   [1.1413 ms 1.1437 ms 1.1462 ms]
                        change: [-0.7358% -0.0798% +0.5611%] (p = 0.82 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  6 (6.00%) high severe

Benchmarking es/parser/jquery
Benchmarking es/parser/jquery: Warming up for 3.0000 s
Benchmarking es/parser/jquery: Collecting 100 samples in estimated 5.1316 s (800 iterations)
Benchmarking es/parser/jquery: Analyzing
es/parser/jquery        time:   [6.3543 ms 6.3626 ms 6.3726 ms]
                        change: [-0.0329% +0.2377% +0.4963%] (p = 0.08 > 0.05)
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  5 (5.00%) high mild
  8 (8.00%) high severe

Benchmarking es/parser/jquery mobile
Benchmarking es/parser/jquery mobile: Warming up for 3.0000 s
Benchmarking es/parser/jquery mobile: Collecting 100 samples in estimated 5.0411 s (500 iterations)
Benchmarking es/parser/jquery mobile: Analyzing
es/parser/jquery mobile time:   [9.9685 ms 9.9926 ms 10.021 ms]
                        change: [-0.1623% +0.1781% +0.5086%] (p = 0.32 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  5 (5.00%) high mild
  6 (6.00%) high severe

Benchmarking es/parser/mootools
Benchmarking es/parser/mootools: Warming up for 3.0000 s
Benchmarking es/parser/mootools: Collecting 100 samples in estimated 5.3106 s (1100 iterations)
Benchmarking es/parser/mootools: Analyzing
es/parser/mootools      time:   [4.8081 ms 4.8163 ms 4.8256 ms]
                        change: [-0.2294% +0.0436% +0.3084%] (p = 0.75 > 0.05)
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  7 (7.00%) high mild
  6 (6.00%) high severe

Benchmarking es/parser/underscore
Benchmarking es/parser/underscore: Warming up for 3.0000 s
Benchmarking es/parser/underscore: Collecting 100 samples in estimated 9.9862 s (10k iterations)
Benchmarking es/parser/underscore: Analyzing
es/parser/underscore    time:   [979.95 us 981.43 us 983.02 us]
                        change: [-0.1476% +0.3311% +0.7981%] (p = 0.18 > 0.05)
                        No change in performance detected.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  5 (5.00%) high severe

Benchmarking es/parser/three
Benchmarking es/parser/three: Warming up for 3.0000 s
Benchmarking es/parser/three: Collecting 100 samples in estimated 5.7091 s (200 iterations)
Benchmarking es/parser/three: Analyzing
es/parser/three         time:   [28.321 ms 28.398 ms 28.484 ms]
                        change: [+0.0223% +0.4348% +0.8602%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 15 outliers among 100 measurements (15.00%)
  9 (9.00%) high mild
  6 (6.00%) high severe

Benchmarking es/parser/yui
Benchmarking es/parser/yui: Warming up for 3.0000 s
Benchmarking es/parser/yui: Collecting 100 samples in estimated 5.4048 s (1100 iterations)
Benchmarking es/parser/yui: Analyzing
es/parser/yui           time:   [4.8431 ms 4.8527 ms 4.8634 ms]
                        change: [-0.5513% -0.2765% +0.0238%] (p = 0.07 > 0.05)
                        No change in performance detected.
Found 16 outliers among 100 measurements (16.00%)
  8 (8.00%) high mild
  8 (8.00%) high severe

```

**Comment by kdy1** [comment]
But the performance of (private) real-world files seems like regressed. I'll investigate

**Review by kdy1 — DISMISSED** [review]
Thanks! I think this will be useful for error recovery in future.

---

swc-bump:
 - swc_ecma_parser

**Comment by kdy1** [comment]
Oh 🤣 it was a draft

**Comment by magic-akari** [comment]
After 3e80402

```
test es/lexer/colors        ... bench:        9605 ns/iter (+/- 22)
test es/lexer/angular       ... bench:     4741204 ns/iter (+/- 10654)
test es/lexer/backbone      ... bench:      653835 ns/iter (+/- 4263)
test es/lexer/jquery        ... bench:     3537659 ns/iter (+/- 7237)
test es/lexer/jquery mobile ... bench:     5677486 ns/iter (+/- 39527)
test es/lexer/mootools      ... bench:     2744882 ns/iter (+/- 17155)
test es/lexer/underscore    ... bench:      543424 ns/iter (+/- 1031)
test es/lexer/three         ... bench:    16011281 ns/iter (+/- 33548)
test es/lexer/yui           ... bench:     2927964 ns/iter (+/- 6808)
```

File: crates/swc_common/src/input.rs [hunk 1]
```diff
@@ -67,6 +67,13 @@ impl<'a> Input for StringInput<'a> {
         self.iter.clone().nth(2).map(|i| i.1)
     }
 
+    #[inline]
+    fn peek_iter(&self) -> Box<dyn Iterator<Item = char> + '_> {
```
Comment (kdy1) [inline crates/swc_common/src/input.rs:71]: This allocates unconditionally

**Review by kdy1 — CHANGES_REQUESTED** [review]
I think `eat` + `reset_to` is better.
I'll profile performance using real-world codes

File: crates/swc_common/src/input.rs (hunk 1 above)
Comment (kdy1) [reply crates/swc_common/src/input.rs:71]: I think we don't need this.
Adding `is_str` is enough, and adding this is a breaking change anyway

Comment (magic-akari) [reply crates/swc_common/src/input.rs:71]: Should I add a `is_str` method to the Input?

Comment (kdy1) [reply crates/swc_common/src/input.rs:71]: Yeah, I think it would be better.

**Review by kdy1 — APPROVED** [review]
Thanks!

**Pull request merged**